onnxruntime/python/tools/quantization/fp16_converter.py

```diff
@@ -0,0 +1,359 @@
+                new_value_info = model.graph.value_info.add()
+                new_value_info.CopyFrom(n)
+                new_value_info.name = output_name
+                new_value_info.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+                # add Cast node (from tensor(float) to tensor(float16) after graph input
+                new_node = [helper.make_node("Cast", [n.name], [output_name], to=10, name=node_name)]
+                model.graph.node.extend(new_node)
+                value_info_list.append(new_value_info)
+                io_casts.add(node_name)
+
+        for i, n in enumerate(model.graph.output):
+            if n.type.tensor_type.elem_type == onnx_proto.TensorProto.FLOAT:
+                input_name = "graph_output_cast_" + str(i)
+                name_mapping[n.name] = input_name
+                graph_io_to_skip.add(n.name)
+
+                node_name = "graph_output_cast" + str(i)
+                # add Cast node (from tensor(float16) to tensor(float) before graph output
+                new_value_info = model.graph.value_info.add()
+                new_value_info.CopyFrom(n)
+                new_value_info.name = input_name
+                new_value_info.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+                new_node = [helper.make_node("Cast", [input_name], [n.name], to=1, name=node_name)]
+                model.graph.node.extend(new_node)
+                value_info_list.append(new_value_info)
+                io_casts.add(node_name)
+
+    while queue:
+        next_level = []
+        for q in queue:
+            # if q is model, push q.graph (GraphProto)
+            if isinstance(q, onnx_proto.ModelProto):
+                next_level.append(q.graph)
+            # if q is model.graph, push q.node.attribute (AttributeProto)
+            if isinstance(q, onnx_proto.GraphProto):
+                for n in q.node:
+                    # if n is in the block list (doesn't support float16), no conversion for the node,
+                    # and save the node for further processing
+                    if n.name in io_casts:
+                        continue
+                    for i in range(len(n.input)):
+                        if n.input[i] in name_mapping:
+                            n.input[i] = name_mapping[n.input[i]]
+                    for i in range(len(n.output)):
+                        if n.output[i] in name_mapping:
+                            n.output[i] = name_mapping[n.output[i]]
+                    # don't add the attr into next_level for the node in node_keep_data_type_list
+                    # so it will not be converted to float16
+                    if n.op_type not in op_allow_list:
+                        node_list.append(n)
+                    else:
+                        if n.op_type == "Cast":
+                            for attr in n.attribute:
+                                if attr.name == "to" and attr.i == 1:
```

Comment:
here 1 and 10, are they DataType enum values? use their names for better readability

Reply:
why change all the cast operators? what if this cast's float output is a compute node that has to use float32?

Reply:
wait, only operators in the allowed list can be here, cast is not in that list. how do you get here?

Reply:
It will add a new cast to node that needs float 32.